core/src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -19,281 +19,590 @@
+                        "or an id for a [" + ScriptType.STORED.getParseField().getPreferredName() + "] script " +
+                        "or [" + ScriptType.FILE.getParseField().getPreferredName() + "] script");
+            }
+
+            if (idOrCode == null) {
+                throw new IllegalArgumentException("must specify an id or code for a script");
+            }
+
+            if (options.size() > 1 || options.size() == 1 && options.get(CONTENT_TYPE_OPTION) == null) {
+                throw new IllegalArgumentException("illegal compiler options [" + options + "] specified");
+            }
+
+            return new Script(type, this.lang == null ? lang : this.lang, idOrCode, options, params);
         }
     }
 
+    private static final ObjectParser<Builder, ParseFieldMatcherSupplier> PARSER = new ObjectParser<>("script", Builder::new);
+
+    static {
+        // Defines the fields necessary to parse a Script as XContent using a ConstructingObjectParser.
+        PARSER.declareField(Builder::setInline, parser -> parser, ScriptType.INLINE.getParseField(), ValueType.OBJECT_OR_STRING);
+        PARSER.declareString(Builder::setStored, ScriptType.STORED.getParseField());
+        PARSER.declareString(Builder::setFile, ScriptType.FILE.getParseField());
+        PARSER.declareString(Builder::setLang, LANG_PARSE_FIELD);
+        PARSER.declareField(Builder::setOptions, XContentParser::mapStrings, OPTIONS_PARSE_FIELD, ValueType.OBJECT);
+        PARSER.declareField(Builder::setParams, XContentParser::map, PARAMS_PARSE_FIELD, ValueType.OBJECT);
+    }
+
     /**
-     * Method for getting the script.
-     * @return The cache key of the script to be compiled/executed.  For dynamic scripts this is the actual
-     *         script source code.  For indexed scripts this is the id used in the request.  For on disk scripts
-     *         this is the file name.
+     * Convenience method to call {@link Script#parse(XContentParser, ParseFieldMatcher, String)}
+     * using the default scripting language.
      */
-    public String getScript() {
-        return script;
+    public static Script parse(XContentParser parser, ParseFieldMatcher matcher) throws IOException {
+        return parse(parser, matcher, DEFAULT_SCRIPT_LANG);
     }
 
     /**
-     * Method for getting the type.
-     *
-     * @return The type of script -- inline, stored, or file.
+     * Convenience method to call {@link Script#parse(XContentParser, ParseFieldMatcher, String)} using the
+     * {@link ParseFieldMatcher} and scripting language provided by the {@link QueryParseContext}.
      */
-    public ScriptType getType() {
-        return type;
+    public static Script parse(XContentParser parser, QueryParseContext context) throws IOException {
+        return parse(parser, context.getParseFieldMatcher(), context.getDefaultScriptLanguage());
```

Comment:
This one is tricky because ScriptSortBuilder uses that specific parse method as part of a ConstructingObjectParser.  I think I'm going to leave these the way they are for now and this can be revisited once the legacy default language is removed.